### ISSUE TITLE
Language server does't offer autocomplete for functions in modules or packages that haven't been opened or loaded

### ISSUE THREAD

**Issue by dekomote** [DetachHead/basedpyright#545]
This happens in vscode. The diagnostic mode is set to "workspace"

`"basedpyright.analysis.diagnosticMode": "workspace"`

I have an example project with 2 packages and 2 modules inside.

<img width="380" alt="image" src="https://github.com/user-attachments/assets/937277d9-7971-49e8-9fd7-34e8822d4feb">

In pack1/mod1.py I have a function

```python
def some_function(arg: str):
    ...
``` 

If I open mod2.py, write "some_" and try to get autocomplete one of two things happen:

1. If mod1 has been opened, or a module that references mod1 or any file in pack1 has been opened, I will get an autocomplete.
 
<img width="851" alt="image" src="https://github.com/user-attachments/assets/6d541d51-e044-4190-9e65-3c74bcf7fda3">

2. If only mod2.py has been opened since opening the editor, I don't get anything.

<img width="586" alt="image" src="https://github.com/user-attachments/assets/2290c993-4ada-4d2c-866c-38ae2e67ce6c">

If I now open pack1/mod1.py, go back to mod2.py and try the same thing, I get an autocomplete.

I have the same setup for sublime and there it works as it should.

<img width="682" alt="image" src="https://github.com/user-attachments/assets/1bdcbe76-dbd6-4777-8ad2-fdd735ceebeb">

Same thing happens with pyright. Hoever, with pylance, the problem doesn't exist.

**Comment by DetachHead** [comment]
thanks for reporting. i've noticed similar behaviour but haven't been able to narrow down anything specific. interesting that it only happens in vscode though

**Comment by DetachHead** [comment]
i reproduced the same issue in sublime text, so i don't think it's specific to vscode

**Comment by dekomote** [comment]
Have you enabled workspace instead of open files only?

**Comment by DetachHead** [comment]
oh wow, i thought `diagnosticMode` was `"workspace"` by default. i should change that (#559)

now that i've set mine to `"workspace"`, i can't seem to reproduce the issue in vscode or sublime text:

![image](https://github.com/user-attachments/assets/9cc9449d-093b-4b8a-b24c-2b090e208640)

how exactly are you configuring `diagnosticMode` in vscode? is it in your workspace or user settings? maybe it's being overwritten or something? also which version of basedpyright are you using?

**Comment by dekomote** [comment]
I use basedpyright.analysis.diagnosticMode, set to workspace.

I use v1.15.2

I tried it again, it doesn't work in vscode. Make sure you haven't opened any files from pack1 prior to opening mod2, even if they are closed now. Once they are read, it works.

**Comment by DetachHead** [comment]
weird, i can reproduce the issue on a different machine. will investigate further, thanks

**Comment by dekomote** [comment]
> is this an issue in pylance?

No. In pylance, it works, even if you set the diagnostic mode to open files only. I think this is what separates pylance from pyright

**Comment by DetachHead** [comment]
ok i think i can reliably reproduce the issue now

> No. In pylance, it works, even if you set the diagnostic mode to open files only. I think this is what separates pylance from pyright

it doesn't seem to work for me in pylance at all. instead i get a quick fix to search for additional imports which i think is a recent change. it does end up showing it but i don't know why they'd add an extra step and make it more annoying:

![image](https://github.com/user-attachments/assets/4b9cbf06-709f-403b-bdcb-22de7a1c1743)

![image](https://github.com/user-attachments/assets/f2c3e13f-9e2c-48ce-82b7-998443fb4697)

issues with import suggestions are always so hard to debug, because the behavior seems to always change randomly.

**Comment by dekomote** [comment]
You probably have "python.languageServer": "None" set in .vscode/settings.json automatically, so pylance doesn't work.
I can reliably get autocomplete with pylance.

**Comment by DetachHead** [comment]
nah i have a separate isolated pylance profile with `python.languageServer` set to `"Default"`

**Comment by rbhanot4739** [comment]
The auto-complete via code actions behavior is quite unpredictable for me. It seems to work for some of the stdlib modules like `os`, `itetools` , `subprocess` but not for others like `logging` and `unittest`, and for third party modules it didn't work completely for me. 

Although this is only via code actions, if I start typing, the entries do show up in the auto-completion popup menu. 

For reference I am using Neovim with native lsp.

**Comment by npip99** [comment]
Confirm the same issue on VSCode and vim for me.

Example file: `test.py`
```
class Test(BaseModel):
    id: str
    test: str

t = Test(
    id="Hey",
    test="yo",
)
```

Startup: https://pastebin.com/RVYHKwDE
Code Action Issue: https://pastebin.com/RcWuYcTf

**Comment by Wizzerinus** [comment]
Actually I've also ran into this issue recently and didn't figure out what's wrong with it. Sometimes a file would properly get saved into autocomplete but an adjacent file in that folder would not. This explains a bit about it but honestly I don't think it explains my issue...

**Comment by zmm2tysu** [comment]
same issue

**Comment by Runderoy** [comment]
same in vim, zed, vs code

**Comment by alexander-chemeza** [comment]
Same with neovim. Strange thing, because Sqlalchemy it supports, but not FastAPI, Django, DRF, Flask.

**Comment by dekomote** [comment]
I've been poking and prodding the code for the past two days. Initially, all the files are analyzed and the symbols are correctly mapped and processed, but somehow, the symbol table gets reset after that.

I'll investigate more.

**Comment by coredumperror** [comment]
Did you ever make any progress on figuring this one out? I'm using the Cursor fork of basedpyright and getting very similar behavior, so I'd love to see this get fixed.

**Comment by dekomote** [comment]
Wasted a ton of time on this issue, and found nothing. All the files in the workspace are getting processed when you open the project correctly, just not offered for autocomplete for some reason I can't understand.

**Comment by coredumperror** [comment]
Makes me think @DetachHead (or someone similarly deeply familiar with how basedpyright works) is going to have to dig into this, then.

**Comment by alexander-chemeza** [comment]
No, I just updated my configuration to the latest version of lspconfig. I figured out that it is looking for some dynamic libraries like fastapi if you insert some its imports in workspace. For example if I write in workspace from FastAPI import FastAPI and from fastapi.security import ... it is looking in these modules.> Did you ever make any progress on figuring this one out? I'm using the Cursor fork of basedpyright and getting very similar behavior, so I'd love to see this get fixed.

**Comment by vhbui02** [comment]
I'm having the same issue. Gosh this is a little troublesome.

**Comment by coredumperror** [comment]
Any chance that this issue could be looked at by someone who can fix it? It's really a major problem.

**Comment by joaoheusi** [comment]
Facing the same issue in Cursor (version 2.2.44) / VSCode version 1.105.1 using either basedpyright or Cursor Pyright
OS: macOS 

Autocomplete only works after opening the module containing the function. If the module hasn’t been opened yet in the session, no suggestions appear.

This significantly impacts usability in larger projects since symbols from unopened modules are not discoverable.
